core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java

```diff
@@ -265,6 +265,10 @@ protected Query doToQuery(QueryParseContext parseContext) throws IOException {
         addBooleanClauses(parseContext, booleanQuery, shouldClauses, BooleanClause.Occur.SHOULD);
         addBooleanClauses(parseContext, booleanQuery, filterClauses, BooleanClause.Occur.FILTER);
 
+        if (booleanQuery.clauses().isEmpty()) {
+            return new MatchAllDocsQuery();
+        }
```

Comment:
I wanted to return as soon as possible but seeing that otherwise we only loop over empty collection, that should not waste much cycles. The creation of the BoolQuery would be unnecessary, but I guess if it makes the code easier to read I can change that.